nimbus/src/enrollment.rs

```diff
@@ -85,90 +365,156 @@ pub fn get_enrollments(db: &Database) -> Result<Vec<EnrolledExperiment>> {
     Ok(result)
 }
 
-/// Update all enrollments. Typically used immediately after the list of
-/// experiments has been refreshed, so some might mean new enrollments, some
-/// might have expired, etc.
-pub fn update_enrollments(
-    db: &Database,
-    mut writer: &mut Writer,
-    nimbus_id: &Uuid,
-    aru: &AvailableRandomizationUnits,
-    app_context: &AppContext,
-) -> Result<()> {
-    log::info!("updating enrollments...");
-    // We might have enrollments for experiments which no longer exist, so we
-    // first build a set of all IDs in both groups.
-    let mut all_slugs = HashSet::new();
-    let experiments = db
-        .get_store(StoreId::Experiments)
-        .collect_all::<Experiment>(writer)?;
+pub(crate) struct EnrollmentsMerger<'a> {
+    nimbus_id: &'a Uuid,
+    available_randomization_units: &'a AvailableRandomizationUnits,
+    app_context: &'a AppContext,
+}
+
+impl<'a> EnrollmentsMerger<'a> {
+    pub(crate) fn new(
+        nimbus_id: &'a Uuid,
+        available_randomization_units: &'a AvailableRandomizationUnits,
+        app_context: &'a AppContext,
+    ) -> Self {
+        Self {
+            nimbus_id,
+            available_randomization_units,
+            app_context,
+        }
+    }
+
+    /// Convenient wrapper around `evolve_enrollments` that fetches the current state of experiments,
+    /// enrollments and user participation from the database.
+    pub(crate) fn evolve_enrollments_in_db(
+        &self,
+        db: &Database,
+        writer: &mut Writer,
```

Comment:
To check my understanding - does the use of `writer` here mean this all runs in the context of a single transaction?

Reply:
Yes I believe so!